services-geocoding/src/main/java/com/mapbox/api/geocoding/v5/models/GeocodingTypeAdapter.java

```diff
@@ -0,0 +1,120 @@
+package com.mapbox.api.geocoding.v5.models;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.gson.Gson;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *
+ * @since 4.9.0
+ */
+abstract class GeocodingTypeAdapter<T> extends TypeAdapter<T> {
+
+  volatile TypeAdapter<String> stringAdapter;
+  final Gson gson;
+
+  GeocodingTypeAdapter(Gson gson) {
+    this.gson = gson;
+  }
+
+  void writeString(@NonNull String name, String value,
+                           @NonNull JsonWriter jsonWriter) throws IOException {
+    jsonWriter.name(name);
+    if (value == null) {
+      jsonWriter.nullValue();
+    } else {
+      TypeAdapter<String> stringAdapter = this.stringAdapter;
+      if (stringAdapter == null) {
+        this.stringAdapter = stringAdapter;
+        stringAdapter = gson.getAdapter(String.class);
+      }
+      stringAdapter.write(jsonWriter, value);
+    }
+  }
+
+  void writeLocalizedMap(@NonNull String name, Map<String, String> localizedMap,
+                                 @NonNull JsonWriter jsonWriter) throws IOException {
+    if (localizedMap != null) {
+      TypeAdapter<String> stringAdapter = this.stringAdapter;
+      if (stringAdapter == null) {
+        this.stringAdapter = stringAdapter;
+        stringAdapter = gson.getAdapter(String.class);
+      }
+      for (String key : localizedMap.keySet()) {
+        jsonWriter.name(name + '_' + key);
+        stringAdapter.write(jsonWriter, localizedMap.get(key));
+      }
+    }
+  }
+
+  void writeLanguages(List<String> languages,
+                              @NonNull JsonWriter jsonWriter) throws IOException {
+    if (languages != null) {
+      TypeAdapter<String> stringAdapter = this.stringAdapter;
+      if (stringAdapter == null) {
+        this.stringAdapter = stringAdapter;
+        stringAdapter = gson.getAdapter(String.class);
+      }
+      for (String lang : languages) {
+        jsonWriter.name("languages_" + lang);
+        stringAdapter.write(jsonWriter, lang);
+      }
+    }
+  }
+
+  String readString(@NonNull JsonReader jsonReader) throws IOException {
+    TypeAdapter<String> stringAdapter = this.stringAdapter;
+    if (stringAdapter == null) {
+      this.stringAdapter = stringAdapter;
+      stringAdapter = gson.getAdapter(String.class);
+    }
+    return stringAdapter.read(jsonReader);
+  }
+
+  @Nullable
+  List<String> readLanguages(@Nullable List<String> languages,
```

Comment:
The passed collections references seem to always be `null`s and are initialized only here, do we need to pass them in the first place then?

Reply:
I believe `readLanguages` will be called multiple times. In case of CarmenContext  json could look like:

Reply:
Will itt be called multiple times with the same list as an argument?

Reply:
Yes. JsonReader is parsing each key/value as they come in. We cannot get all the `language_*`  pairs (and similarly for other pairs and JsonWriter). I believe TypeAdapters are more efficient then JsonSerializer/Deserializer as do not have an intermediate parsed state. Because of that though it is a bit more restrictive.